### PR TITLE
Elemental Analysis Plotting Wrapper/Interface

### LINKED ISSUE #23260: Elemental Analysis - Plotting
Aims to add a plotting wrapper for the production and manipulation of plot widgets.

This will be tied into the other Elemental Analysis functions (loading etc.) in a later branch.

### PR DESCRIPTION
**Pull request by ewancook**
**Description of work.**
Adds a plotting window (interface) that:
- allows 'dynamic' adding and automatic resizing of 1-4 plots
- has a `ComboBox` for plot selection (when changing plot limits)
- has x and y limit changing
- exposes methods to:
    - add vertical lines
    - add horizontal lines
    - plot workspaces (with/without errors)
    - add/remove subplots
    - ~add moveable lines~ (future PR)
- is unit tested (AxisChanger)
- ~is unit tested (Plotting)~ (#23325)

*The Periodic Table functionality has not changed*

 ---
**Notes**
When `Errors` is checked, all the subplots must be cleared then redrawn as `matplotlib` exposes no way to remove a plot from a subplot. If a title/lines were added, then these should be redrawn after `clear()` is called. For this, the additions must be 'saved' to be replayed when the subplots are being redrawn. An additional method is exposed to 'save' arbitrary methods.

The `Add`/`Remove` buttons on the Elemental Analysis window are temporary:
- `Add` will add a plot to the plotting window
- `Remove` will remove a random plot from the plotting window

Once closed, the plotting window can only be reopened by opening a new Elemental Analysis window: this is by design and will change in a future PR

---

**Report to:** Aidy

---
**To test:**
- `Interfaces->Muon->Elemental Analysis`
- All plots should consist of a sample workspace, the result of an addition of two sample workspaces and have a horizontal line at `y = 0.06`, a vertical line at `x = 10100` and a title.
- When plots are added/deleted, the `ComboBox` will update to contain all plots.
- When `Add` is clicked, the limits will stay the same, as the selected plot is the first one in the list.
- All plot elements (title, lines, workspaces, limits) should remain on add/del (unless the plot is deleted)
- Toggling `Errors` will update all plots with error bars, new limits (if necessary), with the plot elements remaining on the plot (this may take a couple of seconds - see *Notes*)
- Typing in new limits (pressing `Return` to update) will update the specified plot with the new limits (the graph can be flipped in the x/y planes if necessary)
- A new plot should automatically have its current limits entered in the axis changer boxes
- An `IndexError` will be thrown if you try and delete a plot where none exist (this is the example delete function in `Elemental Analysis`, else a `KeyError` would be thrown (intentional))
- A `KeyError` will be thrown if a user tries to add a fifth plot

---
Closes #23260

*This does not require release notes* because **they were added in a previous PR**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: scripts/Elemental_Analysis.py [hunk 1]
```diff
@@ -43,6 +66,21 @@ def select_data_file(self):
...
+        self.plotting.plot(name, plot1)
+        plot2 = mantid.Plus(plot1, plot1, OutputWorkspace=str(time()))
+        self.plotting.plot(name, plot2)
+        self.plotting.add_hline(name, 0.06, 0, 1)
+        self.plotting.add_vline(name, 10100, 0, 1)
+
+    def del_plot(self):
+        to_del = random.choice(self.plotting.get_subplots().keys())
+        self.plotting.remove_subplot(to_del)
+
```
Comment (AnthonyLim23) [inline scripts/Elemental_Analysis.py]: not tried it yet but I suspect if you close your GUI main window the plot will hang around. You can stop this by overloading the closeEvent method and manually closing the plot

Comment (ewancook) [reply scripts/Elemental_Analysis.py]: Deliberate for now - I'll be creating a separate issue for this when docking comes about

File: scripts/Muon/GUI/ElementalAnalysis/Plotting/plotting_view.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        """
+
+        def wraps(self, *args, **kwargs):
+            func(self, *args, **kwargs)
+            if len(self.plots):
+                self.figure.tight_layout()
+            self.canvas.draw()
+        return wraps
+
+    def _save_addition(func):
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/ElementalAnalysis/Plotting/plotting_view.py]: where is the code for this and should the ``wraps`` method be inside this?

Comment (ewancook) [reply scripts/Muon/GUI/ElementalAnalysis/Plotting/plotting_view.py]: The `wraps` method should be inside - it allows functions to be [decorated](https://www.python.org/dev/peps/pep-0318/) with `_redo_layout` (`@_redo_layout`).

**Comment by AnthonyLim23** [comment]
Looks good I just have a few comments:

1. I expected the error bar tick box to only apply to the data selected in the combo box. 
2. When a user tries to have a bad number of plots, nothing crashes but I get no feedback either
3. Could we have an all option for the combo box, which sets all of the plots

File: scripts/Muon/GUI/ElementalAnalysis/Plotting/plotting_view.py (hunk 2 above)
Comment (AnthonyLim23) [reply scripts/Muon/GUI/ElementalAnalysis/Plotting/plotting_view.py]: can we have this in a comment (including the link)?

**Comment by AnthonyLim23** [comment]
had a play and a few comments:

1. When a new plot is added it should use the values from all and not its own
2. It is possible to turn the errors on for a single plot then go to all and plot errors is ticked. I think we should untick it if any of the plots do not contain errors. 
3. I can delete workspaces and the plot hangs around. But if I click plot errors this causes an error. We can connect to the delete call for the workspace and have each plot know how many plots it has (then the delete call just removes one). If the count reaches zero we remove the plot. This is best as another PR (after putting it all together)

**Comment by ewancook** [comment]
So `All` is populated from the last selected plot - i.e. if you modify `Plot 3`, then go to `All`, it would have the ranges for Plot 3. The reason for this is that it's likely that you would want to change all plots based on whether the results for one plot were as expected. 

So problems with `1`:
- When the first plot is added, both ranges would be `0` to `0`
- The user won't know the full `x` and `y` limits of the plot if other values are selected: so Aidy's graphs, for example, vary with x limits of `1000` or `8000` and massively different y limits; since you have to add plots sequentially, every plot would have to be subsequently changed to view the full data range in either the x or y direction.
- There's not really such a thing as `All` values, since it updates based on the previously selected plot

`2`: See first point
Seeing as functionality isn't affected (because it was intentional), it's more of a low priority thing that I'd rather do in a later PR - it's quite easy to get around by deselecting then selecting errors, or by selecting a plot without errors, then going to `All`, so is arguably more of a design choice than a problem.

I agree with `3`

**Comment by AnthonyLim23** [comment]
in a later PR is fine, can we just have references to the issues please

**Comment by ewancook** [comment]
Plotting updates as above will be worked on in #23411

**Review by AnthonyLim23 — APPROVED** [review]
I am happy with this and my comments will be addressed in a future issue

**Pull request merged**